opentelemetry/metrics.go

```diff
@@ -3,54 +3,169 @@ package opentelemetry
+	collectorURL, err := url.Parse(agentURL)
+	if err != nil {
+		return nil, err
+	}
+
+	metricOptions := []otlpmetrichttp.Option{
+		otlpmetrichttp.WithURLPath(collectorURL.Path),
+		otlpmetrichttp.WithEndpoint(collectorURL.Host),
+	}
+	if collectorURL.Scheme == "http" {
+		metricOptions = append(metricOptions, otlpmetrichttp.WithInsecure())
+	}
+
+	metricHTTPExporter, err := otlpmetrichttp.New(ctx, metricOptions...)
+	if err != nil {
+		return nil, err
+	}
+
+	httpMetricReader := sdkmetric.NewPeriodicReader(metricHTTPExporter)
+
+	meterProvider := sdkmetric.NewMeterProvider(
+		sdkmetric.WithReader(httpMetricReader),
+		sdkmetric.WithResource(initResource()),
+	)
+
+	return meterProvider, nil
+}
+
+type RelayerMetrics struct {
+	meter metric.Meter
+	Opts  api.MeasurementOption
+
 	DepositEventCount        metric.Int64Counter
+	MessageEventTime         map[string]time.Time
 	ExecutionErrorCount      metric.Int64Counter
 	ExecutionLatency         metric.Int64Histogram
 	ExecutionLatencyPerRoute metric.Int64Histogram
-	BlockDelta               metric.Int64GaugeObserver
-
-	BlockDeltaMap map[uint8]*big.Int
+	BlockDelta               metric.Int64ObservableGauge
+	BlockDeltaMap            map[uint8]*big.Int
 }
 
-// NewChainbridgeMetrics creates an instance of ChainbridgeMetrics
-// with provided OpenTelemetry meter
-func NewChainbridgeMetrics(meter metric.Meter, genericLabels ...attribute.KeyValue) *ChainbridgeMetrics {
-	blockDeltaMap := make(map[uint8]*big.Int)
-	return &ChainbridgeMetrics{
-		DepositEventCount: metric.Must(meter).NewInt64Counter(
-			"chainbridge.DepositEventCount",
-			metric.WithDescription("Number of deposit events per domain"),
-		),
-		ExecutionErrorCount: metric.Must(meter).NewInt64Counter(
-			"chainbridge.ExecutionErrorCount",
-			metric.WithDescription("Number of executions that failed"),
-		),
-		ExecutionLatencyPerRoute: metric.Must(meter).NewInt64Histogram(
-			"chainbridge.ExecutionLatencyPerRoute",
-			metric.WithDescription("Execution time histogram between indexing event and executing it per route"),
-		),
-		ExecutionLatency: metric.Must(meter).NewInt64Histogram(
-			"chainbridge.ExecutionLatency",
-			metric.WithDescription("Execution time histogram between indexing event and executing it"),
-			metric.WithUnit(unit.Milliseconds),
-		),
-		BlockDelta: metric.Must(meter).NewInt64GaugeObserver(
-			"chainbridge.BlockDelta",
-			func(ctx context.Context, result metric.Int64ObserverResult) {
-				for domainID, delta := range blockDeltaMap {
-					labels := append(genericLabels, attribute.Int64("domainID", int64(domainID)))
-					result.Observe(delta.Int64(), labels...)
-				}
-			},
-			metric.WithDescription("Difference between chain head and current indexed block per domain"),
-		),
-		BlockDeltaMap: blockDeltaMap,
+// NewRelayerMetrics initializes OpenTelemetry metrics
+func NewRelayerMetrics(meter metric.Meter, env, relayerID string) (*RelayerMetrics, error) {
```

Comment:
Would prefer to kepp this as labels ...atribute.KeyValue for any possible future labels that we might want to add in the future.
Or whatever the type is now for those

Reply:
But right now it is hardcoded as env and relayerID which is not something I would force someone to use.
Increasing app.go is not a problem IMO if this is the tradeoff

Reply:
Ah, now I get what you mean, yeah you are right, let me think how to make it more extendable so we would not need to change core verytime